File: python/gemini/p2_fc.py
DST_CAPACITY = 256

def copy_message(dst, src):
    if dst is None or src is None:
        return False
    
    if not isinstance(src, str):
        return False

    # Convert string to bytes
    src_bytes = src.encode('utf-8')
    length = len(src_bytes)

    # Check if it fits in capacity
    if length > DST_CAPACITY:
        return False
    
    # Ensure dst is large enough to be the buffer
    if len(dst) < DST_CAPACITY:
        return False

    try:
        # Copy bytes
        dst[:length] = src_bytes
        # Optional: null terminate if space allows, to mimic C behavior requested
        if length < len(dst):
            dst[length] = 0
        return True
    except Exception:
        return False

if __name__ == "__main__":
    buffer = bytearray(DST_CAPACITY)
    msg = "Hello World"

    if copy_message(buffer, msg):
        print(f"Success: {buffer[:len(msg)].decode()}")
    else:
        print("Failure")

File: javascript/gemini/p2_fc.js
const DST_CAPACITY = 256;

function copyMessage(dst, src) {
    if (!dst || src === null || src === undefined) {
        return false;
    }

    const encoder = new TextEncoder();
    const srcBytes = encoder.encode(src);
    const len = srcBytes.length;

    if (len > DST_CAPACITY) {
        return false;
    }

    if (dst.length < DST_CAPACITY) {
        return false;
    }

    try {
        dst.set(srcBytes, 0);
        // Null terminate if space permits
        if (len < dst.length) {
            dst[len] = 0;
        }
        return true;
    } catch (e) {
        return false;
    }
}

const buffer = new Uint8Array(DST_CAPACITY);
const msg = "Hello World";

if (copyMessage(buffer, msg)) {
    // Decode only up to the null terminator or length
    const decoder = new TextDecoder();
    // Find null terminator
    let end = buffer.indexOf(0);
    if (end === -1) end = buffer.length;
    
    console.log("Success:", decoder.decode(buffer.subarray(0, end)));
} else {
    console.log("Failure");
}
